### DISCUSSION TITLE
pushgateway service stop has error log

### DISCUSSION THREAD

**Discussion by jiangpengjul** [prometheus/pushgateway#426] [Q&A]
![image](https://user-images.githubusercontent.com/74517826/135400432-c56c6081-4caa-4b2c-a917-bc35a2619af8.png)

![image](https://user-images.githubusercontent.com/74517826/135400539-c9091e59-52f1-487f-a287-19650b48c53d.png)

**Comment by roidelapluie** [comment]
Can we get the full log line ? The actual error is missing.

Also, we do not provide systemd service file. Can we get the content of your `pushgateway.service` file?

Thanks!

**Reply by jiangpengjul** [reply]
![image](https://user-images.githubusercontent.com/74517826/135425728-e0b17649-ffe5-4094-8ac1-e1844fee2a2c.png)

**Comment by beorn7** [comment]
I guess the log line @jiangpengjul is referring to is `level=error ts=2021-09-30T11:16:55.724Z caller=main.go:199 msg="HTTP server stopped" err="accept tcp [::]:9091: use of closed network connection"` (or similar). Pushgateway indeed emits an error-level log line even on a graceful shutdown. That has to do with the way we handle shutting down the HTTP server. Nothing is really wrong with it, but I guess it could be done in a more elegant way. I would accept a PR. (o:

Relevant code:
* https://github.com/prometheus/pushgateway/blob/079d8d899a0fc47c3b159df2b3a4ffc555170466/main.go#L257
* https://github.com/prometheus/pushgateway/blob/079d8d899a0fc47c3b159df2b3a4ffc555170466/main.go#L199

**Comment by beorn7** [comment]
Note that the Pushgateway still has an exit code of zero upon graceful shutdown. Systemd should be fine with it (and appears to be fine). It's mostly a distraction for users that they see an error where nothing erroneous has happened.

**Comment by beorn7** [comment]
Note that #428 has fixed the behavior.
